### PR TITLE
Fix compact pot protocol

### PR DESCRIPTION
**Pull request by fchirica**
- Fixes whispering of compact PoT between full nodes.
- Retrieves compact PoT done by connected timelord and tries to create compact blocks.

Two more PRs will follow for chain sanitisation:
- timelord <-> vdf_client communication (in chiavdf)
- full_node -> timelord communication (in chia-blockchain).

### CODE REVIEW

File: src/full_node/full_node.py [hunk 1]
```diff
@@ -976,28 +976,32 @@ async def respond_compact_proof_of_time(
...
         )
+        if pot_tuple in self.blockchain.pot_to_chain_height:
+            height = self.blockchain.pot_to_chain_height[pot_tuple]
         if height is None:
+            self.log.info("No block for compact proof of time.")
+            return
+        if not respond_compact_proof_of_time.proof.is_valid(
+            self.constants["DISCRIMINANT_SIZE_BITS"]
+        ):
+            self.log.info("Invalid compact proof of time.")
```
Comment (mariano54) [inline src/full_node/full_node.py:992]: This should be log.error

File: src/full_node/blockchain.py [hunk 2]
```diff
@@ -80,6 +80,8 @@ class Blockchain:
     block_store: BlockStore
     # Coinbase freeze period
     coinbase_freeze: uint32
+    # Map (challenge_hash, iters) to height.
+    pot_to_chain_height: Dict[Tuple[bytes32, uint64], uint32]
```
Comment (mariano54) [inline src/full_node/blockchain.py:84]: It is inconsistent to store this map here, and to also store proof of time heights in full node store

File: src/full_node/blockchain.py [hunk 3]
```diff
@@ -149,6 +152,19 @@ async def _load_chain_from_store(self,) -> None:
         while True:
             self.headers[cur_b.header_hash] = cur_b
             self.height_to_hash[cur_b.height] = cur_b.header_hash
+            full_block: Optional[FullBlock] = await self.block_store.get_block(
```
Comment (mariano54) [inline src/full_node/blockchain.py:155]: This is not efficient. It is reading and parsing every single block in the blockchain, only to obtain the challenge hash and iters. Much better would be to store challenge hash and iters next to the block header, or in a new table

File: src/full_node/blockchain.py [hunk 4]
```diff
@@ -527,6 +543,13 @@ async def receive_block(
 
         # Always immediately add the block to the database, after updating blockchain state
         await self.block_store.add_block(block)
+        if block.proof_of_time is not None:
```
Comment (mariano54) [inline src/full_node/blockchain.py:546]: All blocks should have a PoT here, so instead we should do:
assert block.proof_of_time is not None

File: src/full_node/blockchain.py (hunk 3 above)
Comment (mariano54) [reply src/full_node/blockchain.py:155]: Actually challenge hash and iterations are already stored in the header right? So why do we need to get_block

Comment (mariano54) [reply src/full_node/blockchain.py:155]: Also, this while loop does not include every single block. Between the LCA and the tips, there are a few more blocks added before this loop (line 141). These need to be included in the pot_to_chain_height dict as well.

**Review by mariano54 — CHANGES_REQUESTED** [review]
thanks for the PR, left my comments.

File: src/full_node/blockchain.py (hunk 2 above)
Comment (fchirica) [reply src/full_node/blockchain.py:84]: so this should belong to block_store, right?

File: src/full_node/blockchain.py (hunk 3 above)
Comment (fchirica) [reply src/full_node/blockchain.py:155]: I think PoT belongs to HeaderBlock type, but we have only Header type stored here. I think I'll go for another table to store (challenge_hash, iter, height), and update it on every add_block call.

File: src/full_node/blockchain.py (hunk 2 above)
Comment (mariano54) [reply src/full_node/blockchain.py:84]: Yes that makes sense

File: src/full_node/blockchain.py (hunk 3 above)
Comment (mariano54) [reply src/full_node/blockchain.py:155]: Wait a second, the header has challenge hash and total_iters in it. You can subtract the iters of the previous block to get the iters, you don't need proof of time.

File: src/full_node/blockchain.py [hunk 5]
```diff
@@ -147,7 +149,21 @@ async def _load_chain_from_store(self,) -> None:
...
+                challenge_hash = self.block_store.get_challenge_hash(
+                    cur[i].header_hash
+                )
+                self.block_store.add_proof_of_time(
+                    challenge_hash,
+                    uint64(cur[i].data.total_iters - prev.data.total_iters),
+                    cur[i].data.height,
+                )
+            except Exception as e:
+                log.error(f"Database corruption: {type(e)} {e}")
```
Comment (mariano54) [inline src/full_node/blockchain.py:164]: We should probably let this crash as opposed to catching it.

**Review by mariano54 — CHANGES_REQUESTED** [review]
Also please change the name of the database in the default_config.yaml file. And make sure the DB filename is not copied over from old config files (add it do the DO_NOT_MIGRATE_KEYS dict in init.py migration script).  

Finally, add the changes to the changelog file.

**Review by mariano54 — APPROVED** [review]

**Pull request merged**